korim/src/commonMain/kotlin/com/soywiz/korim/bitmap/BitmapSlice.kt

```diff
@@ -281,68 +393,194 @@ abstract class BmpSlice(
+        it.y = pixelOffsets[1] + pixelOffsets[4] * offsetY + pixelOffsets[5] * offsetX
+    }
 
     fun readPixels(x: Int, y: Int, width: Int, height: Int, out: RgbaArray = RgbaArray(width * height), offset: Int = 0): RgbaArray {
-        check(x in 0 until this.width)
-        check(y in 0 until this.height)
-        check((x + width) in 0 .. this.width)
-        check((y + height) in 0 .. this.height)
+        check(isValidBasePixelPos(x, y))
+        check(isValidBasePixelPos(x + width - 1, y + height - 1))
         check(out.size >= offset + width * height)
-        bmpBase.readPixelsUnsafe(left + x, top + y, width, height, out, offset)
+        readPixelsUnsafe(x, y, width, height, out, offset)
         return out
     }
 
-    fun getRgba(x: Int, y: Int): RGBA = bmpBase.getRgba(left + x, top + y)
-    fun setRgba(x: Int, y: Int, value: RGBA): Unit = bmpBase.setRgba(left + x, top + y, value)
+    fun readPixelsUnsafe(x: Int, y: Int, width: Int, height: Int, out: RgbaArray, offset: Int = 0) {
+        var n = offset
+        val p = PointInt()
+        for (y0 in 0 until height) for (x0 in 0 until width) {
+            if (basePixelPosUnsafe(x0 + x, y0 + y, p) != null) {
+                out[n++] = bmpBase.getRgba(p.x, p.y)
+            } else {
+                out[n++] = Colors.TRANSPARENT_BLACK
+            }
+        }
+    }
+
+    fun getRgba(x: Int, y: Int): RGBA {
+        basePixelPos(x, y)?.let {
+            return bmpBase.getRgba(it.x, it.y)
+        }
+        return Colors.TRANSPARENT_BLACK
+    }
 
-    open fun sliceWithBounds(left: Int, top: Int, right: Int, bottom: Int, name: String? = null): BmpSlice =
-        BitmapSlice(bmp, createRectangleInt(bounds.left, bounds.top, bounds.right, bounds.bottom, left, top, right, bottom), name)
-    open fun sliceWithSize(x: Int, y: Int, width: Int, height: Int, name: String? = null): BmpSlice = sliceWithBounds(x, y, x + width, y + height, name)
-    open fun slice(rect: RectangleInt, name: String? = null): BmpSlice = sliceWithBounds(rect.left, rect.top, rect.right, rect.bottom, name)
-    open fun slice(rect: Rectangle, name: String? = null): BmpSlice = slice(rect.toInt(), name)
+    fun setRgba(x: Int, y: Int, value: RGBA) {
+        basePixelPos(x, y).also {
+            if (it != null) {
+                bmpBase.setRgba(it.x, it.y, value)
+            } else {
+                if (x < 0 || y < 0 || x >= frameWidth || y >= frameHeight) {
+                    throw IllegalArgumentException("Point $x,$y is not in bounds of slice")
+                }
+                bmpBase = extract()
+                bounds.setBoundsTo(0, 0, bmpBase.width, bmpBase.height)
+                virtFrame?.setBoundsTo(0, 0, bmpBase.width, bmpBase.height)
+                bmpCoords = BmpCoordsWithInstanceBase(
+                    SizeInt(bmpBase.width, bmpBase.height),
+                    0f, 0f, 1f, 0f, 1f, 1f, 0f, 1f
```

Comment:
What is this `0f, 0f, 1f, 0f, 1f, 1f, 0f, 1f`

Reply:
`tl_x`, `tl_y`, `tr_x`, ... for a full size slice